### ISSUE TITLE
it's hard for developer to write a common component with vue-demi 

### ISSUE THREAD

**Issue by xuemanchi** [vueuse/vue-demi#153]
【description】:
I want to try create the same component for Vue2 & Vue3 in the same source code, so I find vue-demi. But.. I have seen some projects likes [json-editor-vue](https://github.com/cloydlau/json-editor-vue). For compact Vue2 & Vue3, I found its source code write `render` options for 2 & 3. But i think it's hard for developer to code.

 I prefer to code in `template` in a sfc file. So i have some idea(We can smooth the difference with `Vue2` & `Vue3`  when we compile components). Can we write a component with `setup` & `composition-api` in a `sfc`. And then when we compile the component, we can use a series of plugins for `Vue2`: `unplugin-auto-import`, `unplugin-vue2-script-setup`, `vite-plugin-vue2`, `vue-template-compiler@2.6.x`, and plugins for `Vue3`: `@vitejs/plugin-vue`.   

@antfu Maybe u have thought some idea before create this repo. So can u give me some advice or info?

**Comment by antfu** [comment]
I would suggest directly shipping SFC to npm and let userland plugin to compile it.

**Comment by xuemanchi** [comment]
> I would suggest directly shipping SFC to npm and let userland plugin to compile it.

This is also a way to resolve. I think i may find a way to implement . 
How about create a monorepo,  one package to create SFC, and a package to compile it for `Vue2`, a package to compile it for `Vue3`. :bowtie:

**Comment by wobsoriano** [comment]
Use unbuild for shipping the component directly

**Comment by xuemanchi** [comment]
> 

Thanks! I will use it for my project.

**Comment by rorry121** [comment]
> Use [unbuild](https://github.com/unjs/unbuild) for shipping the component directly

any example?

**Comment by sadeghbarati** [comment]
I have the same problem, **I prefer to code in SFC syntax too**

I'm using [`v-once`](https://sfc.vuejs.org/#eNpdTzFuwzAM/AqhJS1QS7shBCj6DS2pQ7cKIkoQaWcw9PdQcZAhG++OvDtu5rsUuy5oRuN5qrEIMMpSjoECebdTOxBM5XoSVATgz3GFdcg04dE7Bcp691oxXyamkqsM6VTshTNpwNYPw1PgYEZ4MJ3TBh0H8y9SeHSO56nXurDN9c/pZOtCEhNa5DT81nxjrGocTLdogZpG/uRUrLBG7RmwwRnnSNiFTEgCDeaaExzU8dC/etM/9o7UPk27A9sbYxY=) ( head to JS tab on sfc playground ) directive on SFC component which I don't know how to use `v-once` in `.ts` <br/> **(defineComponent)**


> I would suggest directly shipping SFC to npm and let userland plugin to compile it.

- Component must be defined separate field in package.json ( separate from other compiled files )
- User must add `vue-demi` to their dependencies to run component properly

---

### vue 2.6:
![Vue 2 Component.vue](https://user-images.githubusercontent.com/17789047/180299398-4735048a-3efe-4dfa-ad47-a86e5db212d3.png)


### vue 2.7:
![Vue 2 7 Component.vue](https://user-images.githubusercontent.com/17789047/180299377-6d06824b-864e-4924-9052-2693a4687fd2.png)


### vue 3:

![Vue 3 Component.vue](https://user-images.githubusercontent.com/17789047/180299354-de23548b-cb14-49ca-be4f-8e53cbf2b371.png)

**Comment by sadeghbarati** [comment]
> Use [unbuild](https://github.com/unjs/unbuild) for shipping the component directly

unbuild is good with `.ts` defineComponent and `composable`, at this time it's not compatible with `.vue` files

**Comment by sadeghbarati** [comment]
@antfu 

can you create `vue-demi` **starter** that covers everything in vue-demi,  just like https://github.com/antfu/starter-ts ?

I'm using `vite` to bundle Vue package

npm `package alias` seems not working for me

```
"vue": "^3.2.37",
"vue2": "npm:vue@^2.6.14",
"vue2.7": "npm:vue@^2.7.0"
```

vite.config-2.7.ts
```ts
resolve: {
    alias: [
        {
          find: 'vue',
          replacement: resolve(__dirname, './node_modules/vue2.7'),
          // replacement: resolve(__dirname, './node_modules/vue2.7/dist/vue.runtime.common.js'),
          // replacement: 'vue2',
        },
    ],
},
```

tsconfig.naruto.json
```
{
  "compilerOptions": {
    "composite": true,
    "module": "ESNext",
    "moduleResolution": "Node",
    "allowSyntheticDefaultImports": true
    "paths": {
      "vue": [
        "node_modules/vue2.7"
        // "node_modules/vue2.7/dist/vue.runtime.common.js"
        // "node_modules/vue2.7/types/index"
      ],
    }
  },
  "include": ["vite.config-2.7.ts"]
}

```

**Comment by Hisioni** [comment]
have u find the way use `monorepo` to com adapt to use `SFC`?
i try to this but when i use `vite-plugin-dts` generate `d.ts` there still have many compatibility issues.

**Comment by sadeghbarati** [comment]
> have u find the way use `monorepo` to adapt to use `vue-demi` for `SFC`? i try to this but when i use `vite-plugin-dts` generate `d.ts` there still have many compatibility issues.

Hi,

Unfortunately no, I end up using `defineComponent`(.ts) way insteaf of `SFC`(.vue) files

---

for monorepo answer can be yes, but it questions the duty of vue-demi

```
- packages/
  -- vue3
  -- vue2
```

**Comment by Hisioni** [comment]
> > have u find the way use `monorepo` to adapt to use `vue-demi` for `SFC`? i try to this but when i use `vite-plugin-dts` generate `d.ts` there still have many compatibility issues.
> 
> Hi,
> 
> Unfortunately no, I end up using `defineComponent`(.ts) way instead of `SFC`(.vue) files
> 
> for monorepo answer can be yes, but it questions the duty of vue-demi
> 
> ```
> - packages/
>   -- vue3
>   -- vue2
> ```

Yeah, still I am same problem with the questioner, use `TS` render way it's hard to my team, we perfer to `SFC` more, then will use `antfu` says that publish components, use different project `v2`, `v2.7`, `v3` to build it, use git branch to manage it.

**Comment by sadeghbarati** [comment]
this issue also related to this pull-request

https://github.com/vueuse/vue-demi/pull/154

**Comment by sadeghbarati** [comment]
@Hisioni can you share your `vue-demi` repo? If it possible

**Comment by Hisioni** [comment]
> 

Of course, in this [repository](https://github.com/Hisioni/vue-demi-repo-sfc), still have question, i don't know how to config to generate `d.ts` on every vue version

**Comment by Hisioni** [comment]
> @Hisioni can you share your `vue-demi` repo? If it possible

Of course, in this [repository](https://github.com/Hisioni/vue-demi-repo-sfc), still have question, i don't know how to config to generate d.ts on every vue version

**Comment by sadeghbarati** [comment]
> have u find the way use `monorepo` to adapt to use `vue-demi` for `SFC`? i try to this but when i use `vite-plugin-dts` generate `d.ts` there still have many compatibility issues.


@Hisioni I have same issue too, I thought **generating types** only required once with Vue3 as main version in package.json

the error I got when installing vue-demi lib on vue2.7 project
[`Generic type 'DefineComponent' requires between 0 and 11 type arguments.`](https://discord.com/channels/793943652350427136/899997357649821716/1003586803313410118)

found an example for handling type definitions separately :fearful:
https://github.com/alibaba/formily/tree/formily_next/packages/vue

**Comment by Shimada666** [comment]
You can refer to my project, I believe I have explored a relatively elegant solution. It allows you to use the `script setup` syntax and other latest Vue3 syntax to develop a universal component library. The idea is to package different versions together and use the postinstall script to distinguish the user's Vue version and release the corresponding build files.
https://github.com/Shimada666/vue-demi-sfc-component-template

**Comment by sadeghbarati** [comment]
For `vue-demi` components, it's better to use **local registration** over global registration, because of this 👇 

```ts
// global.d.ts or volar.d.ts

declare module '@vue/runtime-core' {  // Vue 3
// declare module 'vue' {   // Vue 2.7
// declare module '@vue/runtime-dom' {  // Vue <= 2.6.14
  export interface GlobalComponents {
    GlobalComponent: typeof import('***')['GlobalComponent']
  }
}

export {}
```

or you can define preset/resolver for [unplugin-vue-components](https://github.com/antfu/unplugin-vue-components#types-for-global-registered-components) in your package

---

- For showing autocomplete props on **Vue3** and **Vue2.7** component check [vue-bridge](https://github.com/vue-bridge/vue-bridge/tree/main/playground) ( how to generate d.ts files )
- Also check this one https://www.npmjs.com/package/@smooth-scrollbar-contrib/vue-test?activeTab=explore based on `vue-bridge` idea

  <details>
  <summary>unbuild config</summary>
  
  ```ts
  import { promises as fsp } from 'fs'
  import { join, resolve } from 'pathe'
  import { definePreset } from 'unbuild'
  
  const clientPreset = definePreset({
    clean: true,
    externals: [
      'vue',
      'vue-demi',
      'smooth-scrollbar',
      '@vueuse/core',
      '@vueuse/shared',
    ],
    rollup: {
      emitCJS: true,
    },
    failOnWarn: false,
  })
  
  const versions = ['v2.6', 'v2.7', 'v3']
  
  async function buildVue() {
    const { build } = await import('unbuild')
    for (const version of versions) {
        await build('.', false, {
          entries: ['src/index'],
          outDir: `dist/${version}`,
          preset: clientPreset,
          declaration: version === 'v3',
          hooks: {
            'build:prepare': async (ctx) => {
              const indexFile = resolve(join(ctx.options.rootDir, 'src/component/index.ts'))
              const versionContent = await fsp.readFile(resolve(join(ctx.options.rootDir, `src/component/${version}.ts`)), 'utf8')
  
              await fsp.writeFile(indexFile, versionContent, 'utf8')
            },
            'build:done': async (ctx) => {
              if (version === 'v2.6' || version === 'v2.7') {
                const dtsFilePath = resolve(join(ctx.options.rootDir, `dist/${version}`, 'index'))
                await fsp.writeFile(`${dtsFilePath}.d.ts`, ["export * from '../src/index'"].join('\n'))
              }
            },
          },
        })
      }
  }
  
  buildVue().catch((err) => {
    console.error(err)
    process.exit(1)
  })

  ```
  </details>


`Vue3`:
```ts
// declaration: true
// dist/index.d.ts
import * as vue_demi from 'vue-demi';

declare const Component: vue_demi.DefineComponent<{...}>;
```

`Vue2.7`:
```ts
// custom index.d.ts
// dist/index.d.ts
// link to source file instead of generating d.ts file

export * from '../src/index'
```

**Comment by andresilva-cc** [comment]
I've been trying to make universal components with Vue and using the hints from here I could make it work. The nice thing is that we don't need to write our template by manually calling the `h` function, which is ugly and hard to maintain, you can use JSX which will compile to the `h-demi` function that @dnldsht created. There are a lot of little details and they are all described on the README of the repo below, so feel free to read and test it for yourself:

https://github.com/andresilva-cc/poc-vue-universal-component

It is not fully polished, it probably lacks some features, but it's working.
